oak-store-document/src/main/java/org/apache/jackrabbit/oak/plugins/document/VersionGarbageCollector.java

```diff
@@ -1160,75 +1172,82 @@ public void collectGarbage(final NodeDocument doc, final GCPhases phases) {
+                            combineInternalPropRemovals(doc, op);
+                            break;
+                        }
+                        case ORPHANS_EMPTYPROPS_UNMERGED_BC : {
+                            collectDeletedProperties(doc, phases, op, traversedState);
+                            collectUnmergedBranchCommits(doc, phases, op, toModifiedMs);
+                            break;
+                        }
+                        case ORPHANS_EMPTYPROPS_BETWEEN_CHECKPOINTS_WITH_UNMERGED_BC : {
+                            collectDeletedProperties(doc, phases, op, traversedState);
+                            collectUnmergedBranchCommits(doc, phases, op, toModifiedMs);
+                            collectRevisionsOlderThan24hAndBetweenCheckpoints(doc, toModifiedMs, phases, op);
+                            break;
+                        }
+                        case ORPHANS_EMPTYPROPS_BETWEEN_CHECKPOINTS_NO_UNMERGED_BC : {
+                            collectDeletedProperties(doc, phases, op, traversedState);
+                            collectRevisionsOlderThan24hAndBetweenCheckpoints(doc, toModifiedMs, phases, op);
+                            break;
+                        }
                     }
                 }
-            }
-            // only add if there are changes for this doc
-            if (op.hasChanges()) {
-                op.equals(MODIFIED_IN_SECS, doc.getModified());
-                garbageDocsCount++;
-                totalGarbageDocsCount++;
-                monitor.info("Collected [{}] garbage count in [{}]", op.getChanges().size(), doc.getId());
-                AUDIT_LOG.info("<Collected> [{}] garbage count in [{}]", op.getChanges().size(), doc.getId());
-                updateOpList.add(op);
-            }
-            if (log.isTraceEnabled() && op.hasChanges()) {
-                // only log in case of changes & debug level enabled
-                log.trace("UpdateOp for [{}] is [{}]", doc.getId(), op);
+                // only add if there are changes for this doc
+                if (op.hasChanges()) {
+                    op.equals(MODIFIED_IN_SECS, doc.getModified());
+                    garbageDocsCount++;
+                    totalGarbageDocsCount++;
+                    monitor.info("Collected [{}] garbage count in [{}]", op.getChanges().size(), doc.getId());
+                    AUDIT_LOG.info("<Collected> [{}] garbage count in [{}]", op.getChanges().size(), doc.getId());
+                    updateOpList.add(op);
+                }
+                if (log.isTraceEnabled() && op.hasChanges()) {
+                    // only log in case of changes & debug level enabled
+                    log.trace("UpdateOp for [{}] is [{}]", doc.getId(), op);
+                }
+            } finally {
+                // pause the fullGC thread to avoid excessive load on the system
+                delayOnModifications(timer.stop().elapsed(MILLISECONDS), cancel, fullGCDelayFactor);
```

Comment:
My initial thought was to delay once per batch (the smaller one, not progressBatch). Agree, doing it on every collectGarbage would also work, but it might be a bit excessive? I'd have done it in `collectFullGC` upon completion of every batch - that would also include the time spent in the querying? Wdyt?

Reply:
I think it makes more sense to do it in a smaller batch, which would also avoid peaks on Mongo. Will make the change.

Reply:
addressed in https://github.com/apache/jackrabbit-oak/pull/1795/commits/a43bd81a1c9421c6aa9427c75eef472c168a0c3b